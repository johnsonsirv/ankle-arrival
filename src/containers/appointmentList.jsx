import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Spinner from 'react-spinkit';
import * as dispatchActions from '../actions';
import AppointmentDetails from '../components/appointmentDetails';
import { CurrentUserType } from '../utils/prop-types';

const mapStateToProps = state => state;

export const AppointmentList = props => {
  const {
    fetchAppointments,
    appointments: { appointments: appointmentData, isFetching },
    currentUser,
  } = props;

  useEffect(() => {
    fetchAppointments(currentUser.token);
  }, [fetchAppointments, currentUser]);

  return (
    <>
      {isFetching && <Spinner name="three-bounce" fadeIn="none" />}
      <div className="appointmentList">
        {appointmentData.map(appointment => (
          <AppointmentDetails
            appointment={appointment}
            key={appointment.id}
          />
        ))}
      </div>
    </>
  );
};

AppointmentList.propTypes = {
  appointments: PropTypes.shape({
    appointments: PropTypes.arrayOf(
      PropTypes.shape({
        dateOfAppointment: PropTypes.string,
        timeOfAppointment: PropTypes.string,
        description: PropTypes.string,
        doctor_firstname: PropTypes.string,
        doctor_lastname: PropTypes.string,
        id: PropTypes.number,
      }).isRequired,
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
  }).isRequired,
  fetchAppointments: PropTypes.func.isRequired,
  currentUser: PropTypes.shape(CurrentUserType).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(AppointmentList);
