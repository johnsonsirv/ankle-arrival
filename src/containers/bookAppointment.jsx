import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Joi from 'joi-browser';
import { toast } from 'react-toastify';
import Spinner from 'react-spinkit';
import DateInput from '../components/forms/inputs/dateInput';
import TimeInput from '../components/forms/inputs/timeInput';
import RichTextField from '../components/forms/inputs/richTextField';
import ReadOnlyTextField from '../components/forms/inputs/readOnlyTextField';
import * as dispatchActions from '../actions';
import { Redirect } from 'react-router-dom';

const mapStateToProps = state => state;

export const BookAppointment = props => {
  const {
    currentUser,
    doctors: { doctors },
    match: { params },
    addNewAppointment,
    appointments: { isFetching, error, newRecord },
  } = props;

  const doctor = doctors.filter(d => d.username === params.username)[0];

  const [booking, setBooking] = useState({
    dateOfAppointment: `${new Date().toLocaleDateString()}`,
    timeOfAppointment: `${new Date().toLocaleTimeString()}`,
    description: '',
    currentUser,
    doctor,
    isValid: false,
  });

  const schema = {
    description: Joi.string().required().label('Description'),
    dateOfAppointment: Joi.string().required().label('Date of Appointment'),
    timeOfAppointment: Joi.string().required().label('Time of Appointment'),
  };

  const validateProperty = ({ name, value }) => {
    const obj = { [name]: value };
    const subSchema = { [name]: schema[name] };
    const { error } = Joi.validate(obj, subSchema);
    return error ? error.details[0].message : null;
  };

  const handleSubmit = e => {
    e.preventDefault();
    const {
      dateOfAppointment,
      timeOfAppointment,
      description,
      doctor,
      currentUser,
    } = booking;

    const params = {
      doctor_id: doctor.id,
      description,
      appointment_date: dateOfAppointment,
      appointment_time: timeOfAppointment,
    };

    addNewAppointment({ params, ...currentUser });
  };

  const handleChange = e => {
    const { value, name } = e.target;
    booking[name] = value;
    const error = validateProperty(e.target);
    booking.isValid = !error;

    setBooking({ ...booking });
    if (!booking.isValid) toast.error(error);
  };

  return (
    <>
      {isFetching && <Spinner name="three-bounce" fadeIn="none" />}
      {newRecord && <Redirect to="/appointments" />}
      {error && error.invalid && (
        <h4>Could not save this record. Try again.</h4>
      )}
      <div>
        <form>
          <ReadOnlyTextField
            name="doctor"
            value={`${doctor.firstname} ${doctor.lastname}`}
          />
          <ReadOnlyTextField
            name="current-user"
            value={`${currentUser.username}`}
          />
          <RichTextField
            name="description"
            value={booking.description}
            onChange={handleChange}
          />
          {/* <DateInput
            onChange={handleChange}
            name="dateOfAppointment"
            value={booking.dateOfAppointment}
          /> */}
          {/* <TimeInput
            onChange={handleChange}
            name="timeOfAppointment"
            value={booking.timeOfAppointment}
          /> */}
          <button
            type="button"
            onClick={handleSubmit}
            id="book-appointment"
            disabled={!booking.isValid}
          >
            Save Appointment
          </button>
        </form>
      </div>
    </>
  );
};

BookAppointment.defaultProps = {
  appointments: null,
};

BookAppointment.propTypes = {
  doctors: PropTypes.shape({
    doctors: PropTypes.arrayOf(
      PropTypes.shape({
        firstname: PropTypes.string,
        lastname: PropTypes.string,
        city: PropTypes.string,
        email: PropTypes.string,
        username: PropTypes.string,
        id: PropTypes.number,
      }).isRequired,
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
  }).isRequired,

  currentUser: PropTypes.shape({
    username: PropTypes.string,
    token: PropTypes.string,
  }).isRequired,
  addNewAppointment: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      username: PropTypes.string,
    }).isRequired,
  }).isRequired,
  appointments: PropTypes.shape({
    isFetching: PropTypes.bool,
    newRecord: PropTypes.bool,
    error: PropTypes.shape({
      invalid: PropTypes.bool,
    }),
  }),
};

export default connect(mapStateToProps, dispatchActions)(BookAppointment);
