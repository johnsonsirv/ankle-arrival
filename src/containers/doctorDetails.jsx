import React from 'react';

const DotorDetails = () => {
  return <div>doctor details page</div>;
};

export default DotorDetails;
