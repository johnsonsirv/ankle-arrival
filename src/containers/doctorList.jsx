import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import Doctor from '../components/doctor';
import * as dispatchActions from '../actions';

const mapStateToProps = state => state;

export const DoctorList = props => {
  const {
    currentUser,
    doctors: { doctors: doctorData, isFetching },
    fetchDoctors,
  } = props;

  useEffect(() => {
    fetchDoctors(currentUser.token);
  }, [fetchDoctors, currentUser.token]);

  return (
    <>
      {isFetching && <Spinner name="three-bounce" fadeIn="none" />}
      {doctorData.map(doctor => (
        <Doctor doctor={doctor} currentUser={currentUser} key={doctor.id} />
      ))}
    </>
  );
};

DoctorList.propTypes = {
  doctors: PropTypes.shape({
    doctors: PropTypes.arrayOf(
      PropTypes.shape({
        firstname: PropTypes.string,
        lastname: PropTypes.string,
        city: PropTypes.string,
        email: PropTypes.string,
        username: PropTypes.string,
        id: PropTypes.number,
      }).isRequired
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
  }).isRequired,

  fetchDoctors: PropTypes.func.isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    token: PropTypes.string,
  }).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(DoctorList);
