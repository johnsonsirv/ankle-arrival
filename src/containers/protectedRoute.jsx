/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import * as dispatchActions from '../actions';

const mapStateToProps = state => state;

const ProtectedRoute = ({
  getCurrentUser,
  isAuthenticated,
  component: Component,
  render,
  ...rest
}) => {
  useEffect(() => {
    getCurrentUser();
  }, []);
  return (
    <Route
      {...rest}
      render={props => {
        if (!isAuthenticated) return <Redirect to="/login" />;
        return Component ? <Component {...props} /> : render(props);
      }}
    />
  );
};

ProtectedRoute.propTypes = {
  getCurrentUser: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, dispatchActions)(ProtectedRoute);
