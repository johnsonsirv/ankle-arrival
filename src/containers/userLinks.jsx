import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import * as dispatchActions from '../actions';
import ProtectedLinks from '../components/protectedLinks';

const mapStateToProps = state => state;

const UserLinks = props => {
  const {
    getCurrentUser,
    currentUser: { isAuthenticated },
  } = props;

  useEffect(() => {
    getCurrentUser();
  }, [getCurrentUser]);

  return (
    <>
      <div>
        {!isAuthenticated && (
          <>
            <NavLink to="/signup">Sign Up|</NavLink>
            <NavLink to="/login">Login|</NavLink>
          </>
        )}
        {isAuthenticated && (
          <>
            <NavLink to="/logout">Logout|</NavLink>
            <ProtectedLinks />
          </>
        )}
      </div>
    </>
  );
};

UserLinks.propTypes = {
  getCurrentUser: PropTypes.func.isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    token: PropTypes.string,
    isAuthenticated: PropTypes.bool,
  }).isRequired,
};

export default connect(mapStateToProps, dispatchActions)(UserLinks);
