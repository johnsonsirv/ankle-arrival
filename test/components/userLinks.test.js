/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable no-undef */
import { UserLinks } from '../../src/containers/userLinks';

function setup() {
  const props = {
    currentUser: {
      isAuthenticated: false,
    },
    getCurrentUser: jest.fn(),
  };
  const enzymeWrapper = shallow(<UserLinks {...props} />);

  return { props, enzymeWrapper };
}

describe('<UserLinks /> rendering', () => {
  const { enzymeWrapper } = setup();

  it('should render correctly', () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });

});
