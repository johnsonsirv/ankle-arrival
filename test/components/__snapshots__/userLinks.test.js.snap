// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserLinks /> rendering should render correctly 1`] = `ShallowWrapper {}`;
